### DISCUSSION TITLE
Proposal: Better Access Control (In Nested Classes)

### DISCUSSION THREAD

**Discussion by SuperS1405** [dotnet/csharplang#6135] [Language Ideas]
In my experience, access control between nesting and nested classes is extremly fiddely;

You can either

1) make the nested class itself private/protected and all its members public, so the nesting class - and it only - has access to the nested class, or

2) make the nested class public and use appropriate access control, so that the nested class has the same access to the nested class as any other.

But what if you need the nested class to be public and yet have members only accessible by the nesting class?

Like in the following case:
I have a generic class List\<T>, which is a circular, strongly linked list.
The nested Node class should be public as it provides mutating methods on a per-node basis, such as Remove(), AddBefore(T) and AddBehind(T).

    // Items marked with @ have to be accessible from the nesting class, List<T>.
    public List<T>
    {
        public class Node
        {
            public T Item { get; set; }
            public List<T> List { get; @ set; }
            public Node Prev { get; @ set; }
            public Node Next { get; @ set;  }

            @ Node(T item, List<T> list, Node prev, Node next) { } 

            public void Delete();
            ...
        }

        private Node root;

        ...
    }

There are workarounds for this problem, but all of them have either unnecessary overhead or are ugly.

1) Use internal access modifier. Gives access to your entire assembly - Great. Why not use public while you are at it? 

2) You can use reflection. Ugly & Inefficient.

3) Playing around with inheritance and encapsulation inmiscellanious ways. Ugly.

4) You can use a nested, private interface. Ugly & doesn't work on constructors and static Members.

So how do we resolve this problem (if I am not the only one who cares about it, that is)?

We could

1) allow the nesting class access to protected members of the nested class.
    Probably the simplest and most elegant solution.

2) introduce a new access modifier, eg. public protected, allowing the nesting class access.
    
  
3) introduce a new access modifier scope(_scope) that gives access to the specified scope,
    eg. 'scope(List\<T>) Node(T item, List\<T> list, Node prev, Node next)'
    and 'public Node Prev {get; scope(List\<T>) set; }'
    Certainly the most complicated to implement, but also the most dynamic,
    giving precise access control in multiply nested classes.

**Comment by HaloFour** [comment]
IIRC all of the nested class behaviors regarding accessibility are inherited from the CLR, so there really isn't a good way of changing that behavior from within C# itself.  There'd need to be a parallel modification to the CLR which would expose/enable further accessibility modes.

**Comment by bondsbw** [comment]
> introduce a new access modifier scope(_scope) that gives access to the specified scope,

If we go this route I think I'd prefer to parameterize the existing access modifiers, and expand it to include namespace scope instead of just class scope:

```c#
protected(namespace_or_type) // Applies to derived types
internal(namespace_or_type) // Applies to types in the same assembly
public(namespace_or_type) // Applies to types in any assembly
```
where _namespace_or_type_ is like:
- `My.Namespace` - Visibility applied to types underneath the namespace hierarchy of _My.Namespace_
- `My.Namespace.MyClass` - Visibility applied to class _MyClass_ in namespace _My.Namespace_ and all types nested inside _MyClass_
- `My.Namespace.MyClass+MyNestedClass` - Visibility applied to nested class _MyNestedClass_ and all types nested inside _MyNestedClass_

Perhaps relative paths could be used as well, which could better support refactoring:

```c#
namespace My.Namespace
{
    public class MyClass
    {
        internal(..) string A; // A is accessible from any type within My.Namespace, including MyClass and MyNestedClass
        public class MyNestedClass
        {
            internal(..) string Q;  // Q is accessible from MyClass and any type nested in MyClass
            internal(...) string R;  // R is accessible from any type underneath My.Namespace
            internal(...MyClass) string S;  // Effectively the same as Q
        }
    }
}
```

I'm using a relative path syntax similar to Python.

_(edited example)_

**Comment by rawlinson** [comment]
How about using a public interface with a corresponding private class?

**Comment by DavidArno** [comment]
As @rawlinson says, there exists a solution to this already: use a public interface and private class:

```cs
public interface IList<T>{}

public interface INode<T>
{
    T Item { get; set; }
    IList<T> List { get; }
    INode<T> Prev { get; }
    INode<T> Next { get; }
}

public class List<T> : IList<T>
{
    private class Node : INode<T>
    {
        public T Item { get; set; }
        public IList<T> List { get; set; }
        public INode<T> Prev { get; set; }
        public INode<T> Next { get; set;  }

        public Node(T item, List<T> list, Node prev, Node next) { } 

        public void Delete() {}
    }

    private INode<T> root;

}
```

**Comment by SuperS1405** [comment]
@rawlinson and @DavidArno: Similar to the privately nested interface workaround, this is a very ugly solution; you have to type everything twice, both in the interface and the implementation, and have no control over static members and constructors (Neither are needed in this particular case, but this was just an exsample for illustration purposes after all).

But if what @HaloFour  said is true it doesn't matter anyways and we'll just have to live with using workarounds.

**Comment by rawlinson** [comment]
> this is a very ugly solution

This does of course depend on the application, but is most cases I would say the separation of interface and implementation is actually beautiful - it has so many benefits.

C# 8 is introducing Default Interface Implementations that would save you some typing, but its use for this purpose would probably be an abuse and indeed be ugly!

**Comment by yaakov-h** [comment]
Surely that is a suboptimal solution, given the recent emphasis in the language on high-performance code? You've got additional allocations and/or boxing.

Though personally, I'd be perfectly happy with using `internal`, and this seems to be the solution used by the BCL as well.

**Comment by DavidArno** [comment]
@SuperS1405,
> this is a very ugly solution

That's clearly a matter of opinion. In my view, creating a public interface and keeping the class internal or private is 100% the correct thing to do. Further, Neither the idea of exposing the class through an interface, nor using internal to expose it to the rest of the assembly are ugly or work-arounds in my experience. What tend to be ugly hacks are the various proposals for even more, unnecessary, access levels.

If you need more than private, internal and public, you are doing it wrong in my opinion.

@yaakov-h,
> Surely that is a suboptimal solution, given the recent emphasis in the language on high-performance code?

Thankfully some parts of the language haven't yet been corrupted by that emphasis and remain focused on good software engineering principles, such as defining a public interface and a private/internal class.

**Comment by Joe4evr** [comment]
> C# 8 is introducing Default Interface Implementations that would save you some typing, but its use for this purpose would probably be an abuse and indeed be ugly!

And even then, you still need to create a type to implement the interface before you can instantiate it, which doesn't really make it a solution for the discussion at hand.

Now, one of the side-effects of Default Interface Implementations is that it'll also allow nesting types *inside* an interface, which would still be "typing everything twice", but will improve the locality. [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBEJinANjASQDsYIFsBjCAgWwAdcIaITYBLAeyIBoYQA3MmzABPAD4ABAAwACCQEYA3AFgAUNLnyALCtVqJAZhlsSZShBkEAchwAmEADwAVAHxqA3mpneZTy6RoZdxkAcwgYRRlkcMiAXy8fAgAZNmQYZxcZFLSg0JiZeNUfSxt7DJkABSR+XLCIgoTvaztHVxkrCAAPGFr8wsa5IwUANhKW8oBlGCgiWygEW1KIAAo/NgDuS2z0ttxUmE3msra6asOl8qIumABKAeKAXkyrgHd2luX1pk29tM3TiD8TZXbo3XQDU5sfiwCwSABM73sMhAY2ObiKPk8GOK3kMvn8TF69Wi9UKOJ8eOS+3K2yJkRJcXuFKMR1amSqgLpUT6TNxLIubQ63S5DIaemxOLxS1Wxg2W2pu3253GJzOqLZMhBt1yZJ8hViQA==).

I just really like the possibilities that brings. My favorite is providing this to API consumers:
```cs
public interface IConfig
{
    public static IConfig DefaultConfig { get; }

    //....
}
```
And then use it like so:
```cs
public class Service
{
    private readonly IConfig _config;

    //oh yeah, let's use Nullable Reference Types too!
    public Service(IConfig? config = null)
    {
        _config = config ?? IConfig.DefaultConfig;
    }
}
```

**Comment by DavidArno** [comment]
@Joe4evr,
> Now, one of the side-effects of Default Interface Implementations is that it'll also allow nesting types *inside* an interface...

Oh. You do realise that you just made me like the Default Interface Implementations feature, as putting the implementation inside the interface is just such a brilliant idea? Don't know whether to hate you or hug you! 😆

**Comment by jnm2** [comment]
@DavidArno 

> Thankfully some parts of the language haven't yet been corrupted by that emphasis

Any time someone claims that an influence is corrupting, I become skeptical. It sounds to me too much like prescriptivism.

**Comment by sab39** [comment]
Is there a reason for this being closed? It still seems like a valuable feature to me, and reading the thread I don't see anyone from the LDM specifically ruling it out. In my particular use case none of the workarounds are suitable. What I'm trying to do is implement DUs using source generators, to allow a developer to write something like:
```cs
[Choice]
public partial class InventoryState
{
    public string Description { get; private init; } = null!;

    [Choice] public static InitState Init { get; } = new()
    {
        Description = "Initializing",
    };

    [Choice] public static StoredState Stored { get; } = new()
    {
        Description = "In Storage",
    };

    [Choice] public static FetchState Fetch { get; } = new()
    {
        Description = "Fetch Requested",
    };
}
```
The source generator would recognize the [Choice] attribute and provide the necessary supporting code, including the public nested subclasses InitState, StoredState and FetchState.

In order to be able to achieve this level of simplicity and clarity in the human written code, I need to be able to use object initializer syntax, so what's exposed needs to be a constructor, not a static delegate. It also needs to be able to target-type the new(), so the generated nested classes must be concrete rather than an interface. But it's also important that their constructors _not_ be available to external code, because the instances generated here should be singletons of their respective types.

I'd like to be able to vote for this proposal, in other words, but in order to do that the ticket would need to be open!

**Reply by sab39** [reply]
Some things I forgot to mention:

The new `file private` accessibility wouldn't help, because the constructors need to be accessible from the human-written part of the code, which is by definition in a separate file from the generated code.

As far as syntax, my suggestion would be to follow the precedent of combining existing access modifiers (a la `private protected` and `protected internal`) and use something like `internal private` for this.

I don't think that lack of underlying CLR support is a problem: the new accessibility is a strict subset of `internal` so it won't be exposed outside the compilation unit anyway, so the compiler can translate to `internal` behind the scenes.
